uvicorn/protocols/http/h11_impl.py

```diff
@@ -454,20 +471,53 @@ async def send(self, message):
 
         elif not self.response_complete:
             # Sending response body
-            if message_type != "http.response.body":
-                msg = "Expected ASGI message 'http.response.body', but got '%s'."
-                raise RuntimeError(msg % message_type)
-
-            body = message.get("body", b"")
-            more_body = message.get("more_body", False)
+            use_sendfile = False
+            if message_type == "http.response.body":
+                body = message.get("body", b"")
+                more_body = message.get("more_body", False)
+            elif (
+                self.allow_sendfile and message_type == "http.response.zerocopysend"
+            ):  # pragma: no cover
+                file_fd = message["file"]
+                sendfile_offset = message.get("offset", None)
+                if sendfile_offset is None:
+                    sendfile_offset = os.lseek(file_fd, 0, os.SEEK_CUR)
+                sendfile_count = message.get("count", None)
+                if sendfile_count is None:
+                    sendfile_count = os.stat(file_fd).st_size - sendfile_offset
+                more_body = message.get("more_body", False)
+                use_sendfile = True
+            else:
+                if self.allow_sendfile:  # pragma: no cover
+                    expect_message_types = (
+                        "http.response.body",
+                        "http.response.zerocopysend",
+                    )
+                else:
+                    expect_message_types = ("http.response.body",)
+                msg = "Expected ASGI message %s, but got '%s'."
+                raise RuntimeError(msg % expect_message_types, message_type)
 
             # Write response body
             if self.scope["method"] == "HEAD":
                 event = h11.Data(data=b"")
+            elif use_sendfile:  # pragma: no cover
+                with os.fdopen(os.dup(file_fd), "rb") as file:
+                    event = SendfileData(file, sendfile_offset, sendfile_count)
+                    for data in self.conn.send_with_data_passthrough(
+                        h11.Data(data=event)
+                    ):
+                        if isinstance(data, SendfileData):
+                            await self.flow.drain()
+                            await self.loop.sendfile(
+                                self.transport, data.file, data.offset, data.count
+                            )
+                        else:
+                            self.transport.write(data)
             else:
                 event = h11.Data(data=body)
-            output = self.conn.send(event)
-            self.transport.write(output)
+                output = self.conn.send(event)
+                self.transport.write(output)
```

Comment:
This looks like it unintentionally changes the behaviour for the `"HEAD"` case.

Reply:
https://github.com/encode/uvicorn/pull/1210/commits/ddb3e58672531cd3034513bd704921c047d2dd7a done.